### PR TITLE
Health server watch method

### PR DESCRIPTION
**Pull request by canguler**

### CODE REVIEW

File: health/health.go [hunk 1]
```diff
@@ -65,13 +61,54 @@ func (s *Server) Check(ctx context.Context, in *healthpb.HealthCheckRequest) (*h
...
+	if _, ok := s.updates[service]; !ok {
+		s.updates[service] = make(map[healthpb.Health_WatchServer]chan healthpb.HealthCheckResponse_ServingStatus)
+	}
+	s.updates[service][stream] = update
+	s.mu.Unlock()
+	for {
+		select {
+		// Status updated. Sends the up-to-date status to the client.
+		case status := <-update:
+			stream.Send(&healthpb.HealthCheckResponse{Status: status})
```
Comment (menghanl) [inline health/health.go]: Check the error returned by send.

File: health/health.go [hunk 2]
```diff
@@ -36,25 +36,21 @@ type Server struct {
...
+		statusMap: map[string]healthpb.HealthCheckResponse_ServingStatus{"": healthpb.HealthCheckResponse_SERVING},
+		updates:   make(map[string]map[healthpb.Health_WatchServer]chan healthpb.HealthCheckResponse_ServingStatus),
 	}
 }
 
 // Check implements `service Health`.
 func (s *Server) Check(ctx context.Context, in *healthpb.HealthCheckRequest) (*healthpb.HealthCheckResponse, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	if in.Service == "" {
```
Comment (menghanl) [inline health/health.go]: Why is this deleted?

File: test/end2end_test.go [hunk 3]
```diff
@@ -2432,6 +2432,102 @@ func testHealthCheckOff(t *testing.T, e env) {
...
+	if err != nil {
+		t.Fatalf("error: %v", err)
+	}
+
+	healthWatchChecker(t, stream12, healthpb.HealthCheckResponse_SERVING)
+
+	hs.SetServingStatus(service2, healthpb.HealthCheckResponse_SERVING)
+	hs.SetServingStatus(service1, healthpb.HealthCheckResponse_NOT_SERVING)
+
+	healthWatchChecker(t, stream11, healthpb.HealthCheckResponse_NOT_SERVING)
```
Comment (menghanl) [inline test/end2end_test.go]: Also do
```go
healthWatchChecker(t, stream12, healthpb.HealthCheckResponse_NOT_SERVING)
```
?

This is to check the update to `service1` would affect both streams watching `service1`.

File: health/health.go (hunk 2 above)
Comment (canguler) [reply health/health.go]: @dfawley realized this. We were always returning SERVING for "". This shouldn't be the case.

File: test/end2end_test.go [hunk 4]
```diff
@@ -2432,6 +2432,101 @@ func testHealthCheckOff(t *testing.T, e env) {
...
+	for _, e := range listTestEnv() {
+		testHealthWatch(t, e)
+	}
+}
+
+func testHealthWatch(t *testing.T, e env) {
+	te := newTest(t, e)
+	hs := health.NewServer()
+
+	service0 := ""
```
Comment (dfawley) [inline test/end2end_test.go]: Could you split this into 2-3 different test functions for the various different things that are being tested by this one?  E.g. `TestHealthWatch_DefaultServiceStartsServing`, `TestHealthWatch_UnknownServiceStartsUnknown`, `TestHealthWatch_MultipleWatchers`?

File: health/health.go [hunk 5]
```diff
@@ -65,13 +61,59 @@ func (s *Server) Check(ctx context.Context, in *healthpb.HealthCheckRequest) (*h
...
 
 // SetServingStatus is called when need to reset the serving status of a service
 // or insert a new service entry into the statusMap.
-func (s *Server) SetServingStatus(service string, status healthpb.HealthCheckResponse_ServingStatus) {
+func (s *Server) SetServingStatus(service string, servingStatus healthpb.HealthCheckResponse_ServingStatus) {
 	s.mu.Lock()
-	s.statusMap[service] = status
+	s.statusMap[service] = servingStatus
+	for _, update := range s.updates[service] {
+		// Clears previous updates, that are not sent to the client, from the channel.
```
[the remark this answers is not in the document]
Comment (canguler) [reply health/health.go]: Added the comment.

File: test/end2end_test.go [hunk 6]
```diff
@@ -2432,6 +2432,211 @@ func testHealthCheckOff(t *testing.T, e env) {
...
+func healthWatchChecker(t *testing.T, stream healthpb.Health_WatchClient, expectedServingStatus healthpb.HealthCheckResponse_ServingStatus) {
+	response, err := stream.Recv()
+	if err != nil {
+		t.Fatalf("error on %v.Recv(): %v", stream, err)
+	}
+	if response.Status != expectedServingStatus {
+		t.Fatalf("response.Status is %v (%v expected)", response.Status, expectedServingStatus)
+	}
+}
 func TestUnknownHandler(t *testing.T) {
```
Comment (dfawley) [inline test/end2end_test.go]: Please keep a blank line between functions.
